### ISSUE TITLE
single_snippet_selection reference does not get removed when referenced snippet is removed.

### ISSUE THREAD

**Issue by Jelmer7** [sulu/sulu#6541]
| Q | A
| --- | ---
| Sulu Version | 2.4.1
| PHP Version | 8.0.14
| DB Version | 8
| Browser Version | Any

# Actual Behavior

https://sulu-io.slack.com/archives/C0430GGCV/p1649939602133329

When a single_snippet_selection is used to refer to a snippet, and the snippet itself is removed the reference keeps existing. This will cause the application to throw an error and crash pages
`[2022-04-14T15:52:18.381110+02:00] request.CRITICAL: Uncaught PHP Exception Sulu\Component\DocumentManager\Exception\DocumentNotFoundException: "Could not find document with ID or path "7f3c136a-42b8-4387-8930-c6a90f8a3a8a"" at /var/www/html/vendor/sulu/sulu/src/Sulu/Component/DocumentManager/NodeManager.php line 54 {"exception":"[object] (Sulu\\Component\\DocumentManager\\Exception\\DocumentNotFoundException(code: 0): Could not find document with ID or path \"7f3c136a-42b8-4387-8930-c6a90f8a3a8a\" at /var/www/html/vendor/sulu/sulu/src/Sulu/Component/DocumentManager/NodeManager.php:54)\n[previous exception] [object] (PHPCR\\ItemNotFoundException(code: 0): Item 7f3c136a-42b8-4387-8930-c6a90f8a3a8a not found in workspace default_live at /var/www/html/vendor/jackalope/jackalope-doctrine-dbal/src/Jackalope/Transport/DoctrineDBAL/Client.php:1640)"} []`

In phpcr_nodes the identifier of the referenced snippet still exists after deleting.

# Expected Behavior

When a referenced snippet is deleted, references to this node should be deleted.

# Steps to Reproduce

- Add `single_snippet_selection` to `pane_text.xml`
- Create snippet
- Add `pane-text` to page
- Select snippet
- Publish page
- Remove snippet in /admin/snippets (snippet itself, not the reference on the pane)
- Visit page

# Possible Solutions

Unkown

**Comment by alexander-schranz** [comment]
That the reference still exist is expected here, until next save is expected. But the exception should not be thrown out and be catched by resolver service or content type.
Can you show the full stacktrace of the error to know which service here need to catch the DocumentNotFoundException. I currently could not find a usage of document manager in the [single snippetselection](https://github.com/sulu/sulu/blob/2.5/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php).

**Comment by luca-rath** [comment]
@alexander-schranz The SnippetResolver is called instead of the DocumentManager, see https://github.com/sulu/sulu/blob/2.5/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php#L111

**Comment by luca-rath** [comment]
## Exception [2/2]: DocumentNotFoundException

```
Sulu\Component\DocumentManager\Exception\DocumentNotFoundException:
Could not find document with ID or path "9b3982d1-a613-4601-9399-8f0b265bcbd9"

  at vendor/sulu/sulu/src/Sulu/Component/DocumentManager/NodeManager.php:54
  at Sulu\Component\DocumentManager\NodeManager->find('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/sulu/sulu/src/Sulu/Component/DocumentManager/Subscriber/Phpcr/FindSubscriber.php:79)
  at Sulu\Component\DocumentManager\Subscriber\Phpcr\FindSubscriber->handleFind(object(FindEvent), 'sulu_document_manager.find', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners(array(object(Closure)), 'sulu_document_manager.find', object(FindEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch(object(FindEvent), 'sulu_document_manager.find')
     (vendor/sulu/sulu/src/Sulu/Component/DocumentManager/DocumentManager.php:40)
  at Sulu\Component\DocumentManager\DocumentManager->find('9b3982d1-a613-4601-9399-8f0b265bcbd9', 'de', array('locale' => null, 'structure_type' => null, 'rehydrate' => true, 'type' => null, 'user' => null, 'force_remove_children' => false, 'omit_modified_domain_event' => false, 'load_ghost_content' => false, 'clear_missing_content' => false, 'ignore_required' => false, 'load_shadow_content' => true, 'auto_name' => true, 'auto_rename' => true, 'auto_name_locale' => 'de', 'path' => null, 'node_name' => null, 'parent_path' => null, 'auto_create' => false, 'override' => false))
     (vendor/sulu/sulu/src/Sulu/Component/Content/Mapper/ContentMapper.php:301)
  at Sulu\Component\Content\Mapper\ContentMapper->load('9b3982d1-a613-4601-9399-8f0b265bcbd9', 'robbreport', 'de')
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Snippet/SnippetResolver.php:50)
  at Sulu\Bundle\SnippetBundle\Snippet\SnippetResolver->resolve(array('9b3982d1-a613-4601-9399-8f0b265bcbd9'), 'robbreport', 'de', null, false)
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php:112)
  at Sulu\Bundle\SnippetBundle\Content\SingleSnippetSelection->resolveSnippet(object(Property))
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php:65)
  at Sulu\Bundle\SnippetBundle\Content\SingleSnippetSelection->getViewData(object(Property))
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Resolver/StructureResolver.php:129)
  at Sulu\Bundle\WebsiteBundle\Resolver\StructureResolver->resolve(object(PageBridge), true)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php:83)
  at Sulu\Bundle\WebsiteBundle\Resolver\ParameterResolver->resolve(array(), object(RequestAnalyzer), object(PageBridge), false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:115)
  at Sulu\Bundle\WebsiteBundle\Controller\WebsiteController->getAttributes(array(), object(PageBridge), false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:65)
  at Sulu\Bundle\WebsiteBundle\Controller\WebsiteController->renderStructure(object(PageBridge), array(), false, false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/DefaultController.php:34)
  at Sulu\Bundle\WebsiteBundle\Controller\DefaultController->indexAction(object(PageBridge), false, false)
     (vendor/symfony/http-kernel/HttpKernel.php:152)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:68)
```

## Exception [1/2]: ItemNotFoundException

```
PHPCR\ItemNotFoundException:
Item 9b3982d1-a613-4601-9399-8f0b265bcbd9 not found in workspace default_live

  at vendor/jackalope/jackalope-doctrine-dbal/src/Jackalope/Transport/DoctrineDBAL/Client.php:1640
  at Jackalope\Transport\DoctrineDBAL\Client->getNodeByIdentifier('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/jackalope/jackalope/src/Jackalope/ObjectManager.php:578)
  at Jackalope\ObjectManager->getNodeByIdentifier('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/jackalope/jackalope/src/Jackalope/Session.php:230)
  at Jackalope\Session->getNodeByIdentifier('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/sulu/sulu/src/Sulu/Bundle/DocumentManagerBundle/Session/Session.php:69)
  at Sulu\Bundle\DocumentManagerBundle\Session\Session->getNodeByIdentifier('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/sulu/sulu/src/Sulu/Bundle/DocumentManagerBundle/Session/Session.php:69)
  at Sulu\Bundle\DocumentManagerBundle\Session\Session->getNodeByIdentifier('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/sulu/sulu/src/Sulu/Component/DocumentManager/NodeManager.php:49)
  at Sulu\Component\DocumentManager\NodeManager->find('9b3982d1-a613-4601-9399-8f0b265bcbd9')
     (vendor/sulu/sulu/src/Sulu/Component/DocumentManager/Subscriber/Phpcr/FindSubscriber.php:79)
  at Sulu\Component\DocumentManager\Subscriber\Phpcr\FindSubscriber->handleFind(object(FindEvent), 'sulu_document_manager.find', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners(array(object(Closure)), 'sulu_document_manager.find', object(FindEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch(object(FindEvent), 'sulu_document_manager.find')
     (vendor/sulu/sulu/src/Sulu/Component/DocumentManager/DocumentManager.php:40)
  at Sulu\Component\DocumentManager\DocumentManager->find('9b3982d1-a613-4601-9399-8f0b265bcbd9', 'de', array('locale' => null, 'structure_type' => null, 'rehydrate' => true, 'type' => null, 'user' => null, 'force_remove_children' => false, 'omit_modified_domain_event' => false, 'load_ghost_content' => false, 'clear_missing_content' => false, 'ignore_required' => false, 'load_shadow_content' => true, 'auto_name' => true, 'auto_rename' => true, 'auto_name_locale' => 'de', 'path' => null, 'node_name' => null, 'parent_path' => null, 'auto_create' => false, 'override' => false))
     (vendor/sulu/sulu/src/Sulu/Component/Content/Mapper/ContentMapper.php:301)
  at Sulu\Component\Content\Mapper\ContentMapper->load('9b3982d1-a613-4601-9399-8f0b265bcbd9', 'robbreport', 'de')
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Snippet/SnippetResolver.php:50)
  at Sulu\Bundle\SnippetBundle\Snippet\SnippetResolver->resolve(array('9b3982d1-a613-4601-9399-8f0b265bcbd9'), 'robbreport', 'de', null, false)
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php:112)
  at Sulu\Bundle\SnippetBundle\Content\SingleSnippetSelection->resolveSnippet(object(Property))
     (vendor/sulu/sulu/src/Sulu/Bundle/SnippetBundle/Content/SingleSnippetSelection.php:65)
  at Sulu\Bundle\SnippetBundle\Content\SingleSnippetSelection->getViewData(object(Property))
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Resolver/StructureResolver.php:129)
  at Sulu\Bundle\WebsiteBundle\Resolver\StructureResolver->resolve(object(PageBridge), true)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Resolver/ParameterResolver.php:83)
  at Sulu\Bundle\WebsiteBundle\Resolver\ParameterResolver->resolve(array(), object(RequestAnalyzer), object(PageBridge), false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:115)
  at Sulu\Bundle\WebsiteBundle\Controller\WebsiteController->getAttributes(array(), object(PageBridge), false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/WebsiteController.php:65)
  at Sulu\Bundle\WebsiteBundle\Controller\WebsiteController->renderStructure(object(PageBridge), array(), false, false)
     (vendor/sulu/sulu/src/Sulu/Bundle/WebsiteBundle/Controller/DefaultController.php:34)
  at Sulu\Bundle\WebsiteBundle\Controller\DefaultController->indexAction(object(PageBridge), false, false)
     (vendor/symfony/http-kernel/HttpKernel.php:152)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:68)
```

**Comment by luca-rath** [comment]
@alexander-schranz I'm gonna take a look at this issue

**Comment by alexander-schranz** [comment]
@luca-rath looks like the exception is not catched here: https://github.com/sulu/sulu/blob/55e8d43d6e52c0e9bee6e6a2813899e74366f12b/src/Sulu/Bundle/SnippetBundle/Snippet/SnippetResolver.php#L50-L54

**Comment by alexander-schranz** [comment]
fixed by #6542 thx for reporting
